elastic/kibana
# Cancel EQL and ES|QL queries when an alerting rule times out from running

## Problem Statement
I noticed when investigating OOM errors in serverless that we have some large EQL queries being returned by Elasticsearch alongside a series of logs mentioning the rule execution timed out. I later on noticed that EQL queries are not cancelled in that scenario, causing task cancellation to not work properly and have Kibana handle responses of tasks it stopped running.

## Solution
To solve this issue, I'm enhancing the ES Client provided to rule type executors so that it makes EQL (and also ES|QL while I'm in the area) cancel queries whenever the rule execution times out by providing the `abortController`'s signal that we trigger when a rule times out.

Additional changes:
- Adding total search duration and es search duration metrics to EQL queries.
- Adding total search duration metrics to ES|QL queries (`took` isn't returned for these types of queries, so we don't know how much ES processing occurred).
- Moving some code in the jest test into a function so my additional tests don't add too much code

## To verify
1. Create an ES|QL alerting rule using ES Query rule type
2. Create an EQL rule in security solution
  - Index pattern: '.kibana-event-log*`
  - EQL: `process where event.kind == "execute"`
3. Notice rules are running successfully
4. Apply the following diff to your codebase. This diff will abort the queries shortly after they are sent to Elasticsearch. This is the easiest approach to ensure we are passing the signal properly through. Unfortunately it is not easy to write queries that would take a while to run for testing purposes nor was it easy to find a mechanism to flood the queue so the test query would timeout to have a full end to end test.
<details>
<summary>Click here to expand</summary>

```
diff --git a/x-pack/plugins/alerting/server/lib/wrap_scoped_cluster_client.ts b/x-pack/plugins/alerting/server/lib/wrap_scoped_cluster_client.ts
index 9ddd22a292b..6ae8e07630e 100644
--- a/x-pack/plugins/alerting/server/lib/wrap_scoped_cluster_client.ts
+++ b/x-pack/plugins/alerting/server/lib/wrap_scoped_cluster_client.ts
@@ -133,10 +133,15 @@ function getWrappedTransportRequestFn(opts: WrapEsClientOpts) {
             opts.rule.spaceId
           } - ${JSON.stringify(params)} - with options ${JSON.stringify(requestOptions)}`
         );
-        const result = (await originalRequestFn.call(opts.esClient.transport, params, {
+
+        const promise = originalRequestFn.call(opts.esClient.transport, params, {
           ...requestOptions,
           signal: opts.abortController.signal,
-        })) as Promise<TResponse> | TransportResult<TResponse, TContext>;
+        });
+
+        opts.abortController.abort();
+
+        const result = (await promise) as Promise<TResponse> | TransportResult<TResponse, TContext>;

         const end = Date.now();
         const durationMs = end - start;
@@ -190,10 +195,16 @@ function getWrappedEqlSearchFn(opts: WrapEsClientOpts) {
           opts.rule.spaceId
         } - ${JSON.stringify(params)} - with options ${JSON.stringify(searchOptions)}`
       );
-      const result = (await originalEqlSearch.call(opts.esClient, params, {
+      const promise = originalEqlSearch.call(opts.esClient, params, {
         ...searchOptions,
         signal: opts.abortController.signal,
-      })) as TransportResult<EqlSearchResponse<TEvent>, unknown> | EqlSearchResponse<TEvent>;
+      });
+
+      opts.abortController.abort();
+
+      const result = (await promise) as
+        | TransportResult<EqlSearchResponse<TEvent>, unknown>
+        | EqlSearchResponse<TEvent>;

       const end = Date.now();
       const durationMs = end - start;
```
</details>

5. Restart the Kibana server (if not done automatically)
6. Use the "Run rule" feature in Stack Management to trigger each created rule to run soon.
7. Wait for the rules to run
8. Notice each rule now fails with the following error message: `search has been aborted due to cancelled execution`. I noticed when writing this test plan that the EQL rule marks as success in the framework but if you "View in App" you will see the error message in security solution.
